// redux-chunk/dist/query.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var query = function query(state, apiCall, args) {
    if (typeof apiCall == 'undefined') {
        throw new Error('The API endpoint does not exist.');
    }

    var actionName = apiCall.actionName;

    state = _extends({}, state);

    var res = null;
    if (state.api[actionName]) {
        if (typeof args != 'undefined') {
            // Select a specific parameter call in the actionName endpoint
            res = state.api[actionName][JSON.stringify(args)];
        } else {
            // Return all requests for the endpoint
            res = state.api[actionName];

            return {
                isLoading: res && !!Object.keys(res).find(function (r) {
                    return res[r].isLoading;
                }),
                hasStarted: res != null,
                hasFailed: res && !!Object.keys(res).find(function (r) {
                    return !!res[r].error;
                }),
                result: res && !res.isLoading ? Object.keys(res).map(function (r) {
                    return {
                        params: JSON.parse(r),
                        payload: res[r].result
                    };
                }) : null,
                headers: res && !res.isLoading ? Object.keys(res).map(function (r) {
                    return res[r].headers;
                }) : {},
                error: res && !res.isLoading ? Object.keys(res).map(function (r) {
                    return res[r].error;
                }) : null,
                status: res && !res.isLoading ? Object.keys(res)[0].status : null
            };
        }
    }

    return _extends({
        isLoading: false,
        hasStarted: res != null,
        hasFailed: res && !!res.error,
        result: null,
        headers: {},
        error: null
    }, res);
};

exports.default = query;
